src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Utf8Formatter/FormattingHelpers.cs

```diff
@@ -92,37 +93,44 @@ public static void WriteDigitsWithGroupSeparator(ulong value, Span<byte> buffer)
         }
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        public static void WriteDigits(uint value, Span<byte> buffer)
+        public static void WriteDigits<TChar>(uint value, Span<TChar> buffer) where TChar : unmanaged, IBinaryInteger<TChar>
         {
-            // We can mutate the 'value' parameter since it's a copy-by-value local.
-            // It'll be used to represent the value left over after each division by 10.
+            Debug.Assert(buffer.Length > 0);
 
             for (int i = buffer.Length - 1; i >= 1; i--)
             {
                 uint temp = '0' + value;
                 value /= 10;
-                buffer[i] = (byte)(temp - (value * 10));
+                buffer[i] = TChar.CreateTruncating(temp - (value * 10));
```

Comment:
It "shouldn't" be an issue, but we have seen some edge cases where the JIT gives up on inlining `CreateTruncating` due to its size.

Would be good to have some basic perf numbers (or disasm checks) showing nothing has regressed for these changes.

If anything has regressed, there are likely some small JIT side fixes we can do to better account for this and avoid the regression